### PR TITLE
Do not update likelihood independent random variables in HMC

### PR DESCRIPTION
**Pull request by emilemathieu**
Issue https://github.com/yebai/Turing.jl/issues/372.

I do not understand why the test `io.jl/save_resume_chain.jl` is failing, when I run it separately locally it passes...

### CODE REVIEW

**Comment by xukai92** [comment]
I also meet the problem of the `io.jl/save_resume_chain.jl` problem before. For some reason, the resume is not stable. Some random seed may break it.

I don't know exactly why but my guess is that as the stage of save and resume, the transformation of vi has some problems...

**Comment by emilemathieu** [comment]
@xukai92 this test is eventually passing (for some obscure reasons) ;)

**Comment by xukai92** [comment]
Cool. Let's keep it in mind and try to figure our why is so fragile later. I will do the review later (tonight or tomorrow)

File: src/core/varinfo.jl [hunk 1]
```diff
@@ -92,10 +96,21 @@ setgid!(vi::VarInfo, gid::Int, vn::VarName) = vi.gids[getidx(vi, vn)] = gid
...
 settrans!(vi::VarInfo, trans::Bool, vn::VarName) = vi.trans[end][getidx(vi, vn)] = trans
 
-getlogp(vi::VarInfo) = vi.logp[end]
-setlogp!(vi::VarInfo, logp::Real) = vi.logp[end] = logp
-acclogp!(vi::VarInfo, logp::Real) = vi.logp[end] += logp
-resetlogp!(vi::VarInfo) = setlogp!(vi, zero(Real))
+getlogprior(vi::VarInfo) = vi.logprior[end]
+setlogprior!(vi::VarInfo, logp::Real) = vi.logprior[end] = logp
+acclogprior!(vi::VarInfo, logp::Real) = vi.logprior[end] += logp
+resetlogprior!(vi::VarInfo) = setlogprior!(vi, zero(Real))
```
Comment (xukai92) [inline src/core/varinfo.jl]: Seems too many levels of interfaces here: as `resetlogprior!` only used in `resetlogp!`, maybe it is not necessary.

File: src/core/varinfo.jl [hunk 2]
```diff
@@ -92,10 +96,21 @@ setgid!(vi::VarInfo, gid::Int, vn::VarName) = vi.gids[getidx(vi, vn)] = gid
...
+
+getloglike(vi::VarInfo) = vi.loglike[end]
+setloglike!(vi::VarInfo, logp::Real) = vi.loglike[end] = logp
+accloglike!(vi::VarInfo, logp::Real) = vi.loglike[end] += logp
+resetloglike!(vi::VarInfo) = setloglike!(vi, zero(Real))
```
Comment (xukai92) [inline src/core/varinfo.jl]: Same here

File: src/core/varinfo.jl [hunk 3]
```diff
@@ -92,10 +96,21 @@ setgid!(vi::VarInfo, gid::Int, vn::VarName) = vi.gids[getidx(vi, vn)] = gid
...
+
+getlogp(vi::VarInfo) = getlogprior(vi) + getloglike(vi)
+getbothlogp(vi::VarInfo) = (getlogprior(vi), getloglike(vi))
+setbothlogp!(vi::VarInfo, bothlogp::Tuple{Real,Real}) = (setlogprior!(vi, bothlogp[1]); setloglike!(vi, bothlogp[2]))
+acclogp!(vi::VarInfo, logp::Real) = accloglike!(vi.loglike, logp) # fallback
+resetlogp!(vi::VarInfo) = (resetlogprior!(vi);resetloglike!(vi))
```
Comment (xukai92) [inline src/core/varinfo.jl]: You miss a space after `;`

File: src/core/varinfo.jl [hunk 4]
```diff
@@ -92,10 +96,21 @@ setgid!(vi::VarInfo, gid::Int, vn::VarName) = vi.gids[getidx(vi, vn)] = gid
...
+acclogprior!(vi::VarInfo, logp::Real) = vi.logprior[end] += logp
+resetlogprior!(vi::VarInfo) = setlogprior!(vi, zero(Real))
+
+getloglike(vi::VarInfo) = vi.loglike[end]
+setloglike!(vi::VarInfo, logp::Real) = vi.loglike[end] = logp
+accloglike!(vi::VarInfo, logp::Real) = vi.loglike[end] += logp
+resetloglike!(vi::VarInfo) = setloglike!(vi, zero(Real))
+
+getlogp(vi::VarInfo) = getlogprior(vi) + getloglike(vi)
+getbothlogp(vi::VarInfo) = (getlogprior(vi), getloglike(vi))
```
Comment (xukai92) [inline src/core/varinfo.jl]: Maybe name it as `getlogptuple`?

File: src/samplers/support/hmc_core.jl [hunk 5]
```diff
@@ -36,13 +46,15 @@ leapfrog(_θ::Union{Vector,SubArray}, p::Vector{Float64}, τ::Int, ϵ::Float64,
...
     grad = gradient(vi, model, spl)
-    verifygrad(grad) || (vi[spl] = θ_old; setlogp!(vi, old_logp); θ = θ_old; p = p_old; break)
+    verifygrad(grad) || (vi[spl] = θ_old; setbothlogp!(vi, old_logp); θ = θ_old; p = p_old; break)
 
     p -= ϵ * grad / 2
 
     τ_valid += 1
   end
 
+  θ[indep_rvs] = θ_indep_rvs
```
Comment (xukai92) [inline src/samplers/support/hmc_core.jl]: The way we use `get_indep_rvs` still do the computation to those independent random variables here... Do you think we can use it outside this function and avoid doing the computation?

**Comment by emilemathieu** [comment]
Thanks @xukai92 for your feedbacks. Have you seen my comment https://github.com/yebai/Turing.jl/pull/395 ?

**Comment by xukai92** [comment]
Sorry I didn't catch. Which comment do you mean?

**Comment by xukai92** [comment]
Maybe you mean https://github.com/yebai/Turing.jl/issues/372#issuecomment-348550809?

**Comment by emilemathieu** [comment]
Indeed sorry

**Comment by yebai** [comment]
Closed because it does not work, see https://github.com/yebai/Turing.jl/issues/372#issuecomment-348550809

**Pull request closed without merging**